rust/datafusion/src/logicalplan.rs

```diff
@@ -1042,6 +998,35 @@ pub fn can_coerce_from(type_into: &DataType, type_from: &DataType) -> bool {
     }
 }
 
+/// A registry of functions used to plan queries programmatically
```

Comment:
I wonder what benefit we get from having Registry be its own `struct` and copying the vec around.

What do you think about something like using a traint. Something like 

```
pub trait UDFFactory {
    pub fn udf(&self, name: &str, args: Vec<Expr>) -> Result<Expr>;
}
```

And then move the implementation of `udf` directly into ExecutionContext.

Reply:
That is an excellent idea. I am trying to implement it, but I am facing a challenge; could you help me here?

I've implemented the trait and moved the implementation for the `ExecutionContextState`, which is where the scalar is and what the DataFrame has access to:

```
impl UDFFactory for ExecutionContextState {
    fn udf(&self, name: &str, args: Vec<Expr>) -> Result<Expr> {
    ...
    }
```

On the dataframe, we will need to do something like 

```
fn udfs(&self) -> &dyn UDFFactory {
        self.ctx_state ...
    }
```

but `self.ctx_state` is under a mutex: `Arc<Mutex<ExecutionContextState>>`.

This seems fair (safety wise): another thread could be trying to register a new UDF while we are trying to use one, in which case one of them should wait.

However, I am struggling to write the interface to handle this.

Reply:
I wonder if you could implement UDFFactory on `DataFrameImpl` as well -- something like (untested)

```
impl UDFFactory for DataFrameImpl {
    fn udf(&self, name: &str, args: Vec<Expr>) -> Result<Expr> {
    self.ctx_state.lock().expect("locked the mutex").udf(name, args)
    }
```

I think this is another good example of why it would be helpful to remove the the Mutex from ExecutionContextState. I'll keep working on that -- I apologize for my slowness this week. I'll be back full time next week and hopefully make faster progress